### DISCUSSION TITLE
`URIError: Failed to decode param`  with Remix Express Server BFF

### DISCUSSION THREAD

**Discussion by ligiaandreica** [remix-run/remix#6576] [Q&A] [answered]
Hi!

Is there a way to handle remix URI errors gracefully and not expose so much information regarding the undelaying express server (which might be considered a security concern)? 

Not sure there's something I can add to handle it since it's coming from `node_modules/express`.

<img width="1023" alt="image" src="https://github.com/remix-run/remix/assets/59609285/767fcc7e-a554-4f32-9fb8-1c55b83f9791">

<br>
<br>

### Context:

I've created a basic [remix app](https://github.com/ligiaandreica/remix-validation/tree/main), using `npx create-remix@latest`:

```
? What type of app do you want to create? Just the basics
? Where do you want to deploy? Choose Remix App Server if you're unsure; it's easy to change deployment targets. Express Server
? TypeScript or JavaScript? TypeScript
? Do you want me to run `npm install`? Yes
```

I use it as [BFF](https://remix.run/docs/en/main/guides/bff) and I've created a basic [page](https://github.com/ligiaandreica/remix-validation/blob/main/app/routes/api.data.ts) that returns data correctly.

### Steps to reproduce:
1. clone https://github.com/ligiaandreica/remix-validation
2. run `npm i`
3. run `npm run build`
4. access `http://localhost:3000/api/data` from browser -> all good returning `{"hello":"world"}` as expected
5. try to access `http://localhost:3000/api/%c0` ->  400 Bad Request,`URIError: Failed to decode param` error
- alternatively, you can use the curl request
```
curl --request GET \
  --url http://localhost:3000/api/%c0
```
```
URIError: Failed to decode param '/api/%c0'
    at decodeURIComponent (<anonymous>)
    at decode_param (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/layer.js:172:12)
    at Layer.match (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/layer.js:123:27)
    at matchLayer (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/index.js:585:18)
    at next (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/index.js:226:15)
    at logger (/Users/ligia/dev/remix-validation/node_modules/morgan/index.js:144:5)
    at Layer.handle [as handle_request] (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/index.js:328:13)
    at /Users/ligia/dev/remix-validation/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/Users/ligia/dev/remix-validation/node_modules/express/lib/router/index.js:346:12)
```

**Comment by kiliman** [comment, marked as answer]
In production (`NODE_ENV=production`), any errors will simply return status **500** and _Unexpected Server Error_. By default, it will log the error on the server via `console.error`. The client will **not** get any error details.

If you want to log your errors to something like a bug tracker, you can use the new `handleError` export for _entry.server.tsx_ introduced in v1.17.0

https://remix.run/docs/en/1.17.0/file-conventions/entry.server#handleerror

**Reply by ligiaandreica** [reply]
Yes, you are totally right! Using `npm run start` command sets the environment to production and the error doesn't get returned to the client. I was thrown by seeing the verbose error in the observability logs.

Thank you very much for your quick help! ❤️
